lib/iaas/GcpClient.js

```diff
@@ -0,0 +1,200 @@
+  }
+
+  upload(options, buffer) {
+    return new Promise((resolve, reject) => {
+      function cleanup() {
+        uploadedStream.removeListener('finish', onfinish);
+        uploadedStream.removeListener('error', onerror);
+      }
+
+      function onerror(err) {
+        cleanup();
+        reject(err);
+      }
+
+      function onfinish() {
+        cleanup();
+        resolve(JSON.parse(buffer));
+      }
+
+      let uploadedStream = this.storageClient
+        .bucket(options.container)
+        .file(options.remote)
+        .createWriteStream();
+
+      uploadedStream.on('error', onerror);
+      uploadedStream.on('finish', onfinish);
+      uploadedStream.end(buffer);
+    });
+  }
+
+  uploadJson(container, file, data) {
+    if (arguments.length < 3) {
+      data = file;
+      file = container;
+      container = this.containerName;
+    }
+    return this
+      .upload({
+        container: container,
+        remote: file
+      }, new Buffer(JSON.stringify(data, null, 2), 'utf8'));
+  }
+
+  downloadJson(container, file) {
+    if (arguments.length < 2) {
+      file = container;
+      container = this.containerName;
+    }
+    return this
+      .download({
+        container: container,
+        remote: file
+      })
+      .then((data) => JSON.parse(data))
+      .catchThrow(SyntaxError, new NotFound(`Object '${file}' not found`));
+  }
+
+  deleteSnapshot(snapshotName) {
+    return Promise
+      .try(() => {
+        return this.computeClient
+          .snapshot(snapshotName)
+          .delete()
+          .then(retval => logger.info(`Deleted snapshot ${snapshotName}`, retval));
+      })
+      .catch(BaseCloudClient.providerErrorTypes.Unauthorized, err => {
+        logger.error(err.message);
+        throw new Unauthorized(`Authorization at google compute failed while deleting snapshot ${snapshotName}`);
+      })
+      .catch(BaseCloudClient.providerErrorTypes.Forbidden, err => {
+        logger.error(err.message);
+        throw new Forbidden(`Authentication at google compute failed while deleting snapshot ${snapshotName}`);
+      })
+      .catchThrow(BaseCloudClient.providerErrorTypes.NotFound,
+        new NotFound(`Snapshot ${snapshotName} not found while deleting`));
+  }
+
+  static validateParams(options) {
+    if (!options) {
+      throw new Error('GcpClient can not be instantiated as backup provider config not found');
+    } else {
+      if (!options.projectId || !options.credentials) {
+        throw new Error('GcpClient can not be instantiated as project id or credentials not found in backup provider config');
+      }
+    }
+    return true;
+  }
+
+  static createStorageClient(options) {
+    return GcpStorage({
+      projectId: options.projectId,
+      credentials: options.credentials
+    });
+  }
+
+  static createComputeClient(settings) {
+    return ComputeClient.createComputeClient(settings).value();
```

Comment:
Somehow this pattern of IAAS specific class taking help of an IAAS agnostic factory doesnt seem right. Why not directly put the GCP Compute creation logic within this function? Factory is used generally when the invoker does not know the flavor. But in this case from a specific flavor we are going to the factory, which seems incorrect. I see that this is following Azure way, but i guess even thats not right.

Reply:
Yeah, I have referenced AzureClient for this. We can discuss and I can implement accordingly.

Reply:
As discussed, I will use GCP computeclient directly here and not call computeclient abstraction.